File: iterator/EmployeeDirectReportIterator.java
package iterator;

public class EmployeeDirectReportIterator implements EmployeeHierarchyIterator {

    private final EmployeeHierarchyCollection employeeHierarchyCollection;
    private int currentDirectReportPosition = 0;

    public EmployeeDirectReportIterator(EmployeeHierarchyCollection employeeHierarchyCollection) {
        this.employeeHierarchyCollection = employeeHierarchyCollection;
    }

    @Override
    public boolean hasNext() {
        return currentDirectReportPosition < employeeHierarchyCollection.getEmployees().size();
    }

    @Override
    public Employee getNext() {
        System.out.println("Iterating through a graph DFS / BFS");
        if (!hasNext()) {
            return null;
        }
        return employeeHierarchyCollection.getEmployees().get(currentDirectReportPosition++);
    }
}

File: typescript/patterns/iterator/employee-direct-report-iterator.ts
class EmployeeDirectReportIterator implements EmployeeHierarchyIterator {
  private employeeHierarchyCollection: EmployeeHierarchyCollection;
  private currentDirectReportPosition: number = 0;

  constructor(employeeHierarchyCollection: EmployeeHierarchyCollection) {
    this.employeeHierarchyCollection = employeeHierarchyCollection;
  }

  getNext(): Employee | null {
    console.log("Iterating through a graph DFS / BFS");
    if (!this.hasNext()) return null;
    return this.employeeHierarchyCollection.getEmployees()[
      this.currentDirectReportPosition++
    ];
  }

  hasNext(): boolean {
    return (
      this.currentDirectReportPosition <
      this.employeeHierarchyCollection.getEmployees().length
    );
  }
}
